### DISCUSSION TITLE
How are the SYCL concepts map into CUDA or HIP?

### DISCUSSION THREAD

**Discussion by csccva** [AdaptiveCpp/AdaptiveCpp#953] [Q&A] [answered]
Hello, 

Just curious how are the `buffers` and the `queue` mapped into CUDA? 

When running with CUDA backend are the buffers obtained just by doing `cuda_malloc` and then `cuda_mempcy` ?
What about the queue? is it mapped to a stream?

Cristian

**Comment by illuhad** [comment, marked as answer]
Hi!

The behavior of buffers is described in detail here: https://github.com/OpenSYCL/OpenSYCL/blob/develop/doc/runtime-spec.md
In short, they rely on device USM pointers (1 per device), so yes, on CUDA there would be a lazily executed `cudaMalloc` followed by `cudaMemcpy` if necessary.

For `queue`, Open SYCL maintains a global stream pool. In general, the work that comes in from all SYCL queues is then distributed across the stream pool. So, while ultimately the work is dispatched to a stream, there usually is no clear 1:1 mapping from SYCL queues to streams.
For `in_order` queues, some backends - in particular CUDA and HIP - do by default create dedicated streams. In that case, there is a 1:1 mapping.

**Comment by csccva** [comment]
Thank you for your reply. 

I will check the runtime-spec in more detail.  

Follow up question about the queues. The kernels submitted to the same queue with  have no dependency between them (both working with USM pointers on  an out of order queue) can be executed in the same time. Is this a requirement of the SYCL standard or is it just that it is not guaranteed by the implementation that they will be executed one after another like on a CUDA stream?
And if the former is the case, are there two streams used under the hood to enforces that the kernels execution is not sequential, while in the case of in order queues or dependencies only 1 CUDA stream is used ?

Cristian

**Reply by illuhad** [reply]
The SYCL specification requires that out-of-order queues support out-of-order behavior, but does not mandate when which optimizations exactly will be applied. So you cannot really rely on any particular optimizations.
As I said, work from SYCL out-of-order queues in Open SYCL will be distributed across a global stream pool that is shared by all SYCL queues.
The implementation will try to expose as much parallelism to the backend (e.g. CUDA) as possible, but executing kernels on different streams does not guarantee that kernels will actually be executed concurrently or in a different order - it only allows this. For example, the driver or hardware scheduler could decide that the hardware is better utilized by still executing kernels in-order.
For in-order queues a single, dedicated stream is used.

**Reply by csccva** [reply]
Thanks a lot for the clarifications.
